src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -0,0 +1,248 @@
+                return new TopDocs(totalHits, scoreDocs);
+            }
+            // format scores using following template:
+            // doc_id | magic_number_1
+            // doc_id | magic_number_2
+            // ...
+            // doc_id | magic_number_2
+            // ...
+            // doc_id | magic_number_2
+            // ...
+            // doc_id | magic_number_1
+            List<ScoreDoc> result = new ArrayList<>();
+            result.add(createStartStopElementForHybridSearchResults(delimiterDocId));
+            for (TopDocs topDoc : topDocs) {
+                if (Objects.isNull(topDoc) || Objects.isNull(topDoc.scoreDocs)) {
+                    result.add(createDelimiterElementForHybridSearchResults(delimiterDocId));
+                    continue;
+                }
+                result.add(createDelimiterElementForHybridSearchResults(delimiterDocId));
+                result.addAll(Arrays.asList(topDoc.scoreDocs));
+            }
+            result.add(createStartStopElementForHybridSearchResults(delimiterDocId));
+            scoreDocs = result.stream().map(doc -> new ScoreDoc(doc.doc, doc.score, doc.shardIndex)).toArray(ScoreDoc[]::new);
+        }
+        return new TopDocs(totalHits, scoreDocs);
+    }
+
+    private TotalHits getTotalHits(int trackTotalHitsUpTo, final List<TopDocs> topDocs, final boolean isSingleShard) {
+        final TotalHits.Relation relation = trackTotalHitsUpTo == SearchContext.TRACK_TOTAL_HITS_DISABLED
+            ? TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO
+            : TotalHits.Relation.EQUAL_TO;
+        if (topDocs == null || topDocs.isEmpty()) {
+            return new TotalHits(0, relation);
+        }
+
+        List<ScoreDoc[]> scoreDocs = topDocs.stream()
+            .map(topdDoc -> topdDoc.scoreDocs)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        Set<Integer> uniqueDocIds = new HashSet<>();
+        for (ScoreDoc[] scoreDocsArray : scoreDocs) {
+            uniqueDocIds.addAll(Arrays.stream(scoreDocsArray).map(scoreDoc -> scoreDoc.doc).collect(Collectors.toList()));
+        }
+        long maxTotalHits = uniqueDocIds.size();
+
+        return new TotalHits(maxTotalHits, relation);
+    }
+
+    private float getMaxScore(final List<TopDocs> topDocs) {
+        if (topDocs.isEmpty()) {
+            return 0.0f;
+        } else {
+            return topDocs.stream()
+                .map(docs -> docs.scoreDocs.length == 0 ? new ScoreDoc(-1, 0.0f) : docs.scoreDocs[0])
+                .map(scoreDoc -> scoreDoc.score)
+                .max(Float::compare)
+                .get();
+        }
+    }
+
+    private DocValueFormat[] getSortValueFormats(final SortAndFormats sortAndFormats) {
+        return sortAndFormats == null ? null : sortAndFormats.formats;
+    }
+
+    /**
+     * Implementation of the HybridCollector that reuses instance of collector on each even call. This allows caller to
```

Comment:
What is special about even calls?

Reply:
I probably formulate this in unclear way, basically the idea is that on every odd call we will create new collector and on every even call we will return that collector instead of creating a new one.

Reply:
right, but why not reuse each time?

Reply:
For concurrent search scenario it may cause problems. Core uses thread pool of runners to execute the search, in case we use state that state may be picked up and used in search in completely different slice of segment (I think that's the term they are using for blocks that one thread runner is using). To summarize this:

concurrent search - we cannot use state. reduce will be called so manual reduce call and collector's state both are not needed
non concurrent search - reduce must be called manually. it's safe to use state because it's single threaded execution

Reply:
Discussed offline, initial question was more focused on non-concurrent search and if we can reuse collector for multiple calls, not in every second call.

We can, just the actual logic is to make max of two calls in that flow. 
[In here](https://github.com/opensearch-project/OpenSearch/blob/2.12/server/src/main/java/org/opensearch/search/query/QueryPhase.java#L352) collectors are created, and then used for search. And later in our code in aggregation processor we call [collector manager.reduce](https://github.com/martin-gaievski/neural-search/blob/hybrid_query_scorers_with_approximations/src/main/java/org/opensearch/neuralsearch/search/query/HybridAggregationProcessor.java#L71).
For every new search request [new collector manager is created](https://github.com/martin-gaievski/neural-search/blob/hybrid_query_scorers_with_approximations/src/main/java/org/opensearch/neuralsearch/search/query/HybridAggregationProcessor.java#L38) and this flow repeats.